Cython/Compiler/ModuleNode.py

```diff
@@ -2030,6 +2033,52 @@ def generate_richcmp_function(self, scope, code):
         code.putln("}")  # switch
         code.putln("}")
 
+    def generate_binop_function(self, scope, slot, code):
+        func_name = scope.mangle_internal(slot.slot_name)
+        code.putln()
+        preprocessor_guard = slot.preprocessor_guard_code()
+        if preprocessor_guard:
+            code.putln(preprocessor_guard)
+        if scope.directives['c_api_binop_methods']:
+            code.putln('#define %s %s' % (func_name, slot.left_slot.slot_code(scope)))
+        else:
+            if slot.left_slot.signature == TypeSlots.binaryfunc:
+                extra_arg = extra_arg_decl = ''
+            elif slot.left_slot.signature == TypeSlots.ternaryfunc:
+                extra_arg = ', extra_arg'
+                extra_arg_decl = ', PyObject* extra_arg'
+            else:
+                error(entry.pos, "Unexpected type lost signature: %s" % slot)
+            def has_slot_method(method_name):
+                entry = scope.lookup(method_name)
+                return bool(entry and entry.is_special and entry.func_cname)
+            def call_slot_method(method_name, reverse):
+                entry = scope.lookup(method_name)
+                if entry and entry.is_special and entry.func_cname:
+                    return "%s(%s%s)" % (entry.func_cname, "right, left" if reverse else "left, right", extra_arg)
+                else:
+                    super = 'Py_TYPE(right)->tp_base' if reverse else 'Py_TYPE(left)->tp_base'
+                    return ('(%s->tp_as_number && %s->tp_as_number->%s)'
+                            ' ? %s->tp_as_number->%s(left, right %s)'
+                            ' : (Py_INCREF(Py_NotImplemented), Py_NotImplemented)') % (
+                        super, super, slot.slot_name, super, slot.slot_name, extra_arg)
+            code.putln(
+                TempitaUtilityCode.load_cached(
```

Comment:
This should use `.load_as_string(…)[1]`. The concrete specialised code is unlikely to be used a second time.

Reply:
Good point.